File: src/main/java/ru/c47harsis/fs/FileType.java
package ru.c47harsis.fs;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public enum FileType {
    YAML("yaml", "yml"),
    JSON("json");

    private final static Pattern PATTERN = Pattern.compile(".*\\.(.+)$");
    private final String[] extensions;

    FileType(String... extensions) {
        this.extensions = extensions;
    }

    public static FileType of(String filename) {
        String extension = getExtension(filename);
        if (extension == null) {
            return null;
        }
        for (FileType type : FileType.values()) {
            for (String ext : type.extensions) {
                if (ext.equalsIgnoreCase(extension)) {
                    return type;
                }
            }
        }
        return null;
    }

    private static String getExtension(String filename) {
        Matcher matcher = PATTERN.matcher(filename);
        if (!matcher.matches()) {
            return null;
        }
        return matcher.group(1);
    }
}

File: src/main/kotlin/ru/c47harsis/config/FileType.kt
package ru.c47harsis.config

enum class FileType(vararg val extensions: String) {
    YAML("yaml", "yml"), JSON("json");

    companion object {
        private val pattern = ".*\\.(.+)$".toRegex()

        fun of(filename: String): FileType? {
            val extension = getExtension(filename) ?: return null
            return values().find {
                it.extensions.any { ext ->
                    ext.equals(extension, ignoreCase = true)
                }
            }
        }

        private fun getExtension(filename: String) = pattern.find(filename)?.groups?.get(1)?.value
    }
}